src/bin/repl/mod.rs

```diff
@@ -0,0 +1,189 @@
+        self.editor.load_history(&path).ok()
+    }
+
+    fn history_path() -> Option<PathBuf> {
+        let mut history = data_dir()?;
+        history.push("saltwater_history");
+        Some(history)
+    }
+
+    fn process_line(&mut self, line: String) {
+        self.editor.add_history_entry(line.clone());
+
+        let line = line.trim();
+        if line.starts_with(PREFIX) {
+            let name = line.split(' ').next().unwrap();
+
+            match self.commands.get(&name[1..]) {
+                Some(action) => action(self, &line[name.len()..]),
+                None => println!("unknown command '{}'", name),
+            }
+        } else {
+            match self.execute_code(line) {
+                Ok(_) => {}
+                Err(err) => {
+                    // TODO: Proper error reporting
+                    println!("error: {}", err.data);
+                }
+            }
+        }
+    }
+
+    fn execute_code(&mut self, code: &str) -> Result<(), CompileError> {
+        let module = initialize_jit_module();
+
+        let expr = analyze_expr(code)?;
+        let expr_ty = expr.ctype.clone();
+        let decl = wrap_expr(expr);
+        let module = ir::compile(module, vec![decl], false).0?;
+
+        let mut jit = JIT::from(module);
+        jit.finalize();
+        let execute_fun = jit
+            .get_compiled_function("execute")
+            .expect("this is not good.");
+
+        match expr_ty {
+            Type::Long(signed) => {
+                let result = unsafe {
+                    let execute: unsafe extern "C" fn() -> u64 = std::mem::transmute(execute_fun);
+                    execute()
+                };
+                match signed {
+                    true => println!("=> {}", result as i64),
+                    false => println!("=> {}", result),
+                }
+            }
+            // TODO: Implement execution for more types
+            ty => println!("error: expression returns unsupported type: {:?}", ty),
+        };
+        Ok(())
+    }
+}
+
+/// Takes an expression and wraps it into a `execute` function that looks like the following:
+///
+/// ```
+/// <type> execute() {
+///     return <expr>;
+/// }
+/// ```
+fn wrap_expr(expr: Expr) -> Locatable<Declaration> {
+    let fun = hir::Variable {
+        ctype: types::Type::Function(types::FunctionType {
+            return_type: Box::new(expr.ctype.clone()),
+            params: vec![],
+            varargs: false,
+        }),
+        storage_class: data::StorageClass::Extern,
+        qualifiers: Default::default(),
+        id: saltwater::InternedStr::get_or_intern("execute"),
+    };
+
+    let span = expr.location;
+    let return_stmt = span.with(hir::StmtType::Return(Some(expr)));
+    let init = hir::Initializer::FunctionBody(vec![return_stmt]);
+    let decl = hir::Declaration {
+        // FIXME: Currently doesn't work. Just make insert() pub?
+        symbol: fun.insert(),
```

Comment:
Hmm, the reason I make this private was because I was thinking about parsing and didn't want people to accidentally add a new identifier ... but it does seem like that would be a major use case for any programmatic use. I'm fine with making it public :)